packages/shell-api/src/session.spec.ts

```diff
@@ -0,0 +1,325 @@
+      expect(db).to.deep.equal(new Database(mongo, 'test', session));
+      expect(session.getDatabase('test')).to.equal(db); // reuses db
+    });
+    it('advanceOperationTime', () => {
+      const ts = { ts: 1 } as any;
+      session.advanceOperationTime(ts);
+      expect(serviceProviderSession.advanceOperationTime).to.have.been.calledOnceWith(ts);
+    });
+    it('advanceClusterTime', () => {
+      try {
+        session.advanceClusterTime();
+      } catch (e) {
+        return expect(e.name).to.equal('MongoshUnimplementedError');
+      }
+      expect.fail('Error not thrown');
+    });
+    it('endSession', () => {
+      session.endSession();
+      expect(serviceProviderSession.endSession).to.have.been.calledOnceWith();
+    });
+    it('getClusterTime', () => {
+      serviceProviderSession.clusterTime = 100;
+      expect(session.getClusterTime()).to.equal(100);
+    });
+    it('getOperationTime', () => {
+      serviceProviderSession.operationTime = 200;
+      expect(session.getOperationTime()).to.equal(200);
+    });
+    it('hasEnded', () => {
+      serviceProviderSession.hasEnded = 100 as any; // mystery: testing with false makes this error bc of the spy
+      expect(session.hasEnded()).to.equal(100);
+    });
+    it('startTransaction', () => {
+      serviceProviderSession.startTransaction.returns();
+      session.startTransaction({ readPreference: options.readPreference });
+      expect(serviceProviderSession.startTransaction).to.have.been.calledOnceWith({ readPreference: options.readPreference });
+    });
+    it('commitTransaction', () => {
+      serviceProviderSession.commitTransaction.resolves();
+      session.commitTransaction();
+      expect(serviceProviderSession.commitTransaction).to.have.been.calledOnceWith();
+    });
+    it('abortTransaction', () => {
+      serviceProviderSession.abortTransaction.resolves();
+      session.abortTransaction();
+      expect(serviceProviderSession.abortTransaction).to.have.been.calledOnceWith();
+    });
+  });
+  describe('integration', () => {
+    const replId = 'rs0';
+
+    const testServers = startTestCluster(
+      ['--single', '--replSet', replId],
+      ['--single', '--replSet', replId],
+      ['--single', '--replSet', replId],
+      ['--single', '--replSet', replId]
+    );
+    let cfg: {_id: string, members: {_id: number, host: string, priority: number}[]};
+    let serviceProvider: CliServiceProvider;
+    let internalState: ShellInternalState;
+    let mongo: Mongo;
+    let session: Session;
+
+    const delay = util.promisify(setTimeout);
+    const ensureMaster = async(timeout): Promise<void> => {
+      while (!(await mongo.getDB(ADMIN_DB).isMaster()).ismaster) {
+        if (timeout > 32000) {
+          return expect.fail(`Waited for ${cfg.members[0].host} to become master, never happened`);
+        }
+        await delay(timeout);
+        timeout *= 2; // try again but wait double
+      }
+    };
```

Comment:
Do we want to put this into a helper file, to avoid having duplicate code between here and the other replset tests?

Reply:
will do!